uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,146 @@
+import time
+from typing import List, Optional
+
+from uaclient import actions, cli, contract, entitlements, messages
+from uaclient.api import exceptions
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo
+from uaclient.clouds import AutoAttachCloudInstance  # noqa: F401
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    DataObject,
+    Field,
+    IntDataValue,
+    StringDataValue,
+    data_list,
+)
+from uaclient.entitlements.entitlement_status import CanEnableFailure
+
+
+class FullAutoAttachOptions(DataObject):
+    fields = [
+        Field("enable", data_list(StringDataValue), False),
+        Field("enable_beta", data_list(StringDataValue), False),
+        Field("retries", IntDataValue, False),
+    ]
+
+    def __init__(
+        self,
+        enable: Optional[List[str]],
+        enable_beta: Optional[List[str]],
+        retries: Optional[int],
+    ):
+        self.enable = enable
+        self.enable_beta = enable_beta
+        self.retries = retries
+
+
+class FullAutoAttachResult(DataObject, AdditionalInfo):
+    pass
+
+
+def get_auto_attach_token(cfg: UAConfig, cloud: AutoAttachCloudInstance):
+    contract_client = contract.UAContractClient(cfg)
+    try:
+        tokenResponse = contract_client.request_auto_attach_contract_token(
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def is_any_beta(cfg: UAConfig, ents: List[str]) -> bool:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False
+
+
+def full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    retries_limit = 3
+    if options.retries:
+        retries_limit = options.retries
+    already_attached = False  # avoid AlreadyAttachedOnProError after 1st time
+    while 0 < retries_limit:
+        if not already_attached:
+            instance = None  # type: Optional[AutoAttachCloudInstance]
+            token = None  # Optional[str]
+            instance = actions.get_cloud_instance(cfg)
+            actions.detach_before_auto_attach(cfg)
+            allow_enable = not any([options.enable, options.enable_beta])
+            token = get_auto_attach_token(cfg, instance)  # type: ignore
```

Comment:
Maybe I am missing something here, but I believe the retry should only be executed for this function in particular. Additionally, this function is responsible for making two separate requests:

* The cloud metadata
* The contracts server with that metadata

For the cloud metadata, we already have a timeout mechanism in place. Maybe we should do something similar to the contract call directly, instead of looping here in the API code. What do you all think ? @renanrodrigo @CalvoM @orndorffgrant

Reply:
@CalvoM this abstraction is still retrying on any exception that is raised by those calls. I think instead of having the retry loop here, we could use the `util.retry` decorator directly on the functions we are calling. The only downside is that the retry limits will not be a parameter that can be passed to the api, but I think it is still worth it

Reply:
I thought of this approach, and had the same concern. If it is worth it then it is okay.